packages/backend-modules/base/lib/ControlBus.js

```diff
@@ -0,0 +1,202 @@
+const Redis = require('./Redis')
+const { logger } = require('@orbiting/backend-modules-logger')
+
+/**
+ * @typedef {import('redis').RedisClient} RedisClient
+ */
+
+/** @type {RedisClient | undefined} */
+let ControlBusPublisher
+/** @type {RedisClient | undefined} */
+let ControlBusSubscriber
+let isConnected = false
+
+const controlBusLogger = logger.child({}, { msgPrefix: '[ControlBus] ' })
+const baseTopic = 'republik:control_bus:'
+const subscriptions = new Map()
+
+async function connect() {
+  if (isConnected || ControlBusPublisher) {
+    return
+  }
+
+  try {
+    controlBusLogger.debug('Connecting to Redis...')
+    ControlBusPublisher = Redis.connect()
+    ControlBusSubscriber = ControlBusPublisher.duplicate()
+
+    setupListeners()
+
+    isConnected = true
+  } catch (err) {
+    controlBusLogger.error({ error: err }, 'Failed to create Redis clients')
+    if (ControlBusPublisher) ControlBusPublisher.quit()
+    ControlBusPublisher = null
+    ControlBusSubscriber = null
+    isConnected = false
+  }
+}
+
+/**
+ * Gracefully disconnects from Redis.
+ */
+async function disconnect() {
+  if (!isConnected) {
+    return
+  }
+  isConnected = false
+  if (ControlBusSubscriber) {
+    ControlBusSubscriber.quit()
+    ControlBusSubscriber = null
+  }
+  if (ControlBusPublisher) {
+    ControlBusPublisher.quit()
+    ControlBusPublisher = null
+  }
+  controlBusLogger.debug('Disconnected from Redis.')
+}
+
+function publish(topic, message) {
+  if (!isConnected || !ControlBusPublisher) {
+    controlBusLogger.warn(
+      { topic },
+      'Publish called but not connected. Ignoring.',
+    )
+    return Promise.resolve()
+  }
+
+  const fullTopicName = baseTopic + topic
+  return ControlBusPublisher.publish(fullTopicName, JSON.stringify(message))
+}
+
+function subscribe(topic, cb) {
+  if (!isConnected || !ControlBusSubscriber) {
+    controlBusLogger.warn(
+      { topic },
+      'Subscribe called but not connected. Ignoring.',
+    )
+    return () => {}
+  }
+
+  const fullTopicName = baseTopic + topic
+  if (!subscriptions.has(fullTopicName)) {
+    subscriptions.set(fullTopicName, new Set())
+    ControlBusSubscriber.subscribe(fullTopicName, (err) => {
```

Comment:
Isn't this also an async function, e.g. needs await and try/catch instead of a node-style callback?

Reply:
that is indeed a oversight on my part. I tried to make it work with the non bluebird promisyfyed client at first.